### ISSUE TITLE
Error reading openPMD data after #4606

### ISSUE THREAD

**Issue by Theodotus1243** [ComputationalRadiationPhysics/picongpu#4698]
I have installed `picongpu@530a46eb9b4f5b9d4239f2e7f0ea5af2ea24321b` version of `20.09.2023`
But I did it using spack, so I have `adios2@2.8.0` and `openpmd-api@0.14.5`

After commits https://github.com/ComputationalRadiationPhysics/picongpu/pull/4606 https://github.com/ComputationalRadiationPhysics/picongpu/pull/4607 https://github.com/ComputationalRadiationPhysics/picongpu/pull/4608 I can't read openPMD data using `openPMD-reader`
It returns 
```
[ADIOS2] Warning: Attribute with name /data/0/fields/e_all_momentumDensity/geometry has no type in backend.
[ADIOS2] Warning: Attribute with name /data/0/fields/e_all_momentumDensity/geometry has no type in backend.
[~ADIOS2IOHandler] An error occurred: [ADIOS2] Requested attribute (/data/0/fields/e_all_momentumDensity/geometry) not found in backend.
```

I see that it can be related to renaming, `particleMomentumDensity` -> `MomentumDensity`
But I just don't know what package need to be updated to include this infopmation 
Is it latest `openpmd-api` or `adios2` or else.
Here is an example of data I can't read [simData_200000.bp.zip](https://github.com/ComputationalRadiationPhysics/picongpu/files/12765829/simData_200000.bp.zip)
And here is old working example [simData_200000.bp_OLD_WORKING.zip](https://github.com/ComputationalRadiationPhysics/picongpu/files/12765929/simData_200000.bp_OLD_WORKING.zip)

Comes from https://github.com/ComputationalRadiationPhysics/picongpu/issues/3996

**Comment by PrometheusPi** [comment]
Thanks @Theodotus1243 for reporting. 
I am a bit confused by the the `geometry` part in `/data/0/fields/e_all_momentumDensity/geometry. Where is this coming from? That is something that should not be in PIConGPU openPMD output, nor should it be in the standard as far as I know.

**Comment by Theodotus1243** [comment]
@PrometheusPi thanks for responce 
Then maybe i will not need to install `openpmd-api@develop`

Here is my debs
```
==> 51 installed packages
-- linux-ubuntu20.04-zen / gcc@10.3.0 ---------------------------
adios2@2.8.0          libffi@3.4.2          perl@5.34.1
autoconf@2.69         libiconv@1.16         picongpu@develop
automake@1.16.5       libpciaccess@0.16     pkgconf@1.8.0
berkeley-db@18.1.40   libpng@1.6.37         pmix@4.1.2
boost@1.79.0          libsigsegv@2.13       pngwriter@0.7.0
bzip2@1.0.8           libtool@2.4.7         popt@1.16
c-blosc@1.21.1        libxml2@2.9.13        readline@8.1
cmake@3.23.1          lz4@1.9.3             rsync@3.2.4
cuda@11.7.0           m4@1.4.19             snappy@1.1.8
diffutils@3.8         mpark-variant@1.4.0   sz@2.1.12
freetype@2.11.1       ncurses@6.2           toml11@3.7.1
gdbm@1.19             nlohmann-json@3.10.5  util-macros@1.19.3
hdf5@1.12.2           numactl@2.0.14        xxhash@0.8.1
hwloc@2.7.1           openmpi@4.1.3         xz@5.2.5
libedit@3.1-20210216  openpmd-api@0.14.5   zfp@0.5.5
libevent@2.1.12       openssh@9.0p1         zlib@1.2.12
libfabric@1.14.1      openssl@1.1.1o        zstd@1.5.2

```

Here is my code
[param.zip](https://github.com/ComputationalRadiationPhysics/picongpu/files/12769223/param.zip)
[1.cfg.txt](https://github.com/ComputationalRadiationPhysics/picongpu/files/12769232/1.cfg.txt)

And in previous message I've send and exampel openpmd output
OLD is from picongpu `July` version
And NEW from `~20.09.2023`

**Comment by Theodotus1243** [comment]
What I did, I've updated version to latest dev so that I will be able to save `momentum X and Y` at the same time as described here https://github.com/ComputationalRadiationPhysics/picongpu/issues/3996
But when I updated version and made this change
```
/* MomentumDensity X section */
    using MomentumDensity_X_Seq = deriveField::CreateEligible_t<
        VectorAllPhysicalSpecies,
        deriveField::derivedAttributes::MomentumDensity<0>
    >;

    /* MomentumDensity Y section */
    using MomentumDensity_Y_Seq = deriveField::CreateEligible_t<
        VectorAllPhysicalSpecies,
        deriveField::derivedAttributes::MomentumDensity<1>
    >;

    /** FieldTmpSolvers groups all solvers that create data for FieldTmp ******
     *
     * FieldTmpSolvers is used in @see FieldTmp to calculate the exchange size
     */
    using FieldTmpSolvers = MakeSeq_t<
        Density_Seq,
        EnergyDensity_Seq,
        MomentumDensity_X_Seq,
        MomentumDensity_Y_Seq
    >;
```

I've got this error

**Comment by NeonBohdan** [comment]
@PrometheusPi in the previous message there is a weakest part of code. Do I wrote it wrong?

**Comment by Theodotus1243** [comment]
@PrometheusPi in the previous message there is a weakest part of code. Do I wrote it wrong?

**Comment by PrometheusPi** [comment]
@Theodotus1243 I am currently on vacation and will only be able to look at the issue in about a week. 
Perhaps @franzpoeschel can help in the meantime.

**Comment by franzpoeschel** [comment]
Looking at the dataset, it seems like some attributes are written in the wrong places:

```
  float     /data/200000/fields/e_all_momentumDensity/x                             {1, 256, 192} = 0 / 0
  string    /data/200000/fields/e_all_momentumDensity/x/axisLabels                  attr   = {"z", "y", "x"}
  string    /data/200000/fields/e_all_momentumDensity/x/dataOrder                   attr   = "C"
  string    /data/200000/fields/e_all_momentumDensity/x/fieldSmoothing              attr   = "none"
  string    /data/200000/fields/e_all_momentumDensity/x/geometry                    attr   = "cartesian"
  double    /data/200000/fields/e_all_momentumDensity/x/gridGlobalOffset            attr   = {1417.58, 0, 0}
  float     /data/200000/fields/e_all_momentumDensity/x/gridSpacing                 attr   = {14.7665, 14.7665, 14.7665}
  double    /data/200000/fields/e_all_momentumDensity/x/gridUnitSI                  attr   = 0.000151695
  float     /data/200000/fields/e_all_momentumDensity/x/position                    attr   = {0, 0, 0} 
  float     /data/200000/fields/e_all_momentumDensity/x/timeOffset                  attr   = 0
  double    /data/200000/fields/e_all_momentumDensity/x/unitDimension               attr   = {-2, 1, -1, 0, 0, 0, 0}
  double    /data/200000/fields/e_all_momentumDensity/x/unitSI                      attr   = 0.32974
  float     /data/200000/fields/e_all_momentumDensity/y                             {1, 256, 192} = 0 / 0
  string    /data/200000/fields/e_all_momentumDensity/y/axisLabels                  attr   = {"z", "y", "x"}
  string    /data/200000/fields/e_all_momentumDensity/y/dataOrder                   attr   = "C"
  string    /data/200000/fields/e_all_momentumDensity/y/fieldSmoothing              attr   = "none"
  string    /data/200000/fields/e_all_momentumDensity/y/geometry                    attr   = "cartesian"
  double    /data/200000/fields/e_all_momentumDensity/y/gridGlobalOffset            attr   = {1417.58, 0, 0}
  float     /data/200000/fields/e_all_momentumDensity/y/gridSpacing                 attr   = {14.7665, 14.7665, 14.7665}
  double    /data/200000/fields/e_all_momentumDensity/y/gridUnitSI                  attr   = 0.000151695
  float     /data/200000/fields/e_all_momentumDensity/y/position                    attr   = {0, 0, 0} 
  float     /data/200000/fields/e_all_momentumDensity/y/timeOffset                  attr   = 0
  double    /data/200000/fields/e_all_momentumDensity/y/unitDimension               attr   = {-2, 1, -1, 0, 0, 0, 0}
  double    /data/200000/fields/e_all_momentumDensity/y/unitSI                      attr   = 0.32974
```

This would then probably be a write-time error. I will have a look, but likely not before mid of next week.

**Comment by franzpoeschel** [comment]
Note to myself: Error handling also looks too coarse in openPMD-api 0.15 here, only the faulty meshes should be skipped, but it skips all of them:


```
> openpmd-ls simData_200000.bp
openPMD series: simData_200000
openPMD standard: 1.1.0
openPMD extensions: 1

data author: unknown
data created: 2023-09-22 16:41:10 +0300
data backend: ADIOS2
generating machine: unknown
generating software: PIConGPU (version: 0.8.0-dev)
generating software dependencies: unknown

number of iterations: 1 (fileBased)
  all iterations: [ADIOS2] Warning: Attribute with name /data/200000/fields/i_all_momentumDensity/geometry has no type in backend.
Cannot read meshes in iteration 200000 and will skip them due to read error:
Read Error in backend ADIOS2
Object type:    Attribute
Error type:     NotFound
Further description:    /data/200000/fields/i_all_momentumDensity/geometry
200000 

number of meshes: 0

number of particle species: 1
  all particle species:
    probe
```

**Comment by franzpoeschel** [comment]
@psychocoderHPC  PRs #4606 #4607 and #4680 do not use the openPMD-api correctly. Looking into the verbose log of a PIConGPU run, PIConGPU asks openPMD to create a dataset with name `C_all_momentumDensity/x`:

```
[0x2e92248->0x2f5ee08] CREATE_DATASET: C_all_momentumDensity/x
```

.. instead of creating first a mesh `C_all_momentumDensity` and then within that mesh a component `x`. 

The error introduced by the above PRs:

```cpp
// wrong:
auto record_component = iteration.meshes["C_all_momentumDensity/x"][SCALAR]

// right:
auto record_component = iteration.meshes["C_all_momentumDensity"]["x"]
```

With the current implementation, PIConGPU creates folders that contain a slash, and after this things go south.

**Comment by franzpoeschel** [comment]
> Note to myself: Error handling also looks too coarse in openPMD-api 0.15 here, only the faulty meshes should be skipped, but it skips all of them:
> 
> ```
> > openpmd-ls simData_200000.bp
> openPMD series: simData_200000
> openPMD standard: 1.1.0
> openPMD extensions: 1
> 
> data author: unknown
> data created: 2023-09-22 16:41:10 +0300
> data backend: ADIOS2
> generating machine: unknown
> generating software: PIConGPU (version: 0.8.0-dev)
> generating software dependencies: unknown
> 
> number of iterations: 1 (fileBased)
>   all iterations: [ADIOS2] Warning: Attribute with name /data/200000/fields/i_all_momentumDensity/geometry has no type in backend.
> Cannot read meshes in iteration 200000 and will skip them due to read error:
> Read Error in backend ADIOS2
> Object type:    Attribute
> Error type:     NotFound
> Further description:    /data/200000/fields/i_all_momentumDensity/geometry
> 200000 
> 
> number of meshes: 0
> 
> number of particle species: 1
>   all particle species:
>     probe
> ```

Fixed with https://github.com/openPMD/openPMD-api/pull/1535:

```
> bin/openpmd-ls simData_200000.bp
openPMD series: simData_200000
openPMD standard: 1.1.0
openPMD extensions: 1

data author: unknown
data created: 2023-09-22 16:41:10 +0300
data backend: ADIOS2
generating machine: unknown
generating software: PIConGPU (version: 0.8.0-dev)
generating software dependencies: unknown

number of iterations: 1 (fileBased)
  all iterations: [ADIOS2] Warning: Attribute with name /data/200000/fields/i_all_momentumDensity/geometry has no type in backend.
[AbstractIOHandlerImpl] IO Task READ_ATT failed with exception. Clearing IO queue and passing on the exception.
Cannot read mesh with name 'i_all_momentumDensity' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:    Attribute
Error type:     NotFound
Further description:    /data/200000/fields/i_all_momentumDensity/geometry
[ADIOS2] Warning: Attribute with name /data/200000/fields/e_all_momentumDensity/geometry has no type in backend.
[AbstractIOHandlerImpl] IO Task READ_ATT failed with exception. Clearing IO queue and passing on the exception.
Cannot read mesh with name 'e_all_momentumDensity' and will skip it due to read error:
Read Error in backend ADIOS2
Object type:    Attribute
Error type:     NotFound
Further description:    /data/200000/fields/e_all_momentumDensity/geometry
200000 

number of meshes: 7
  all meshes:
    B
    E
    e_all_density
    e_all_energyDensity
    i_all_density
    i_all_energyDensity
    picongpu_idProvider

number of particle species: 1
  all particle species:
    probe
```

This does NOT fix the original bug described in this issue

**Comment by psychocoderHPC** [comment]
> @psychocoderHPC PRs #4606 #4607 and #4680 do not use the openPMD-api correctly. Looking into the verbose log of a PIConGPU run, PIConGPU asks openPMD to create a dataset with name `C_all_momentumDensity/x`:
> 
> ```
> [0x2e92248->0x2f5ee08] CREATE_DATASET: C_all_momentumDensity/x
> ```
> 
> .. instead of creating first a mesh `C_all_momentumDensity` and then within that mesh a component `x`.
> 
> The error introduced by the above PRs:
> 
> ```c++
> // wrong:
> auto record_component = iteration.meshes["C_all_momentumDensity/x"][SCALAR]
> 
> // right:
> auto record_component = iteration.meshes["C_all_momentumDensity"]["x"]
> ```
> 
> With the current implementation, PIConGPU creates folders that contain a slash, and after this things go south.

I think we need to rename the data set to `C_all_momentumDensity_x` because IMO openPMD API is not able to write the dataset as group correctly in case someone is dumping `C_all_momentumDensity/x` and `C_all_momentumDensity/y`, in this case it required that we update the mesh information e.g. component names of `C_all_momentumDensity` because both components will be written independently in PIConGPU and at the moment where one component is written it is not known that for the same mesh we will write later a second component. But even if it will be possible to update mesh information we do not know in PIConGPU that we already wrote a different mesh component before.

IMO we need to keep each output independent. The initial goal of the PRs which broke code was to provide a mesh but it looks like it is not simply possible, but by giving at least a unique name to the dataset we are able to write "moment/x" and "monetum/y` from the same simulation which was before not possible.

**Comment by psychocoderHPC** [comment]
The issue should be fixed with https://github.com/ComputationalRadiationPhysics/picongpu/pull/4703

**Comment by franzpoeschel** [comment]
> The issue should be fixed with #4703

Please consider using this fix instead https://github.com/ComputationalRadiationPhysics/picongpu/pull/4704
That one writes data the way you originally intended
